crates/wasmparser/src/readers/core/types.rs

```diff
@@ -284,6 +284,36 @@ impl RefType {
     /// `i31ref`.
     pub const I31REF: Self = RefType::from_u32(Self::NULLABLE_BIT | Self::I31_TYPE);
 
+    /// A non-nullable untyped function reference aka `(ref func)`.
+    pub const REF_FUNC: Self = RefType::from_u32(Self::FUNC_TYPE);
+
+    /// A non-nullable reference to an extern object aka `(ref extern)`.
+    pub const REF_EXTERN: Self = RefType::from_u32(Self::EXTERN_TYPE);
+
+    /// A non-nullable reference to any object aka `(ref any)`.
+    pub const REF_ANY: Self = RefType::from_u32(Self::ANY_TYPE);
+
+    /// A non-nullable reference to no object aka `(ref none)`.
+    pub const REF_NONE: Self = RefType::from_u32(Self::NONE_TYPE);
+
+    /// A non-nullable reference to a noextern object aka `(ref noextern)`.
+    pub const REF_NOEXTERN: Self = RefType::from_u32(Self::NOEXTERN_TYPE);
+
+    /// A non-nullable reference to a nofunc object aka `(ref nofunc)`.
+    pub const REF_NOFUNC: Self = RefType::from_u32(Self::NOFUNC_TYPE);
+
+    /// A non-nullable reference to an eq object aka `(ref eq)`.
+    pub const REF_EQ: Self = RefType::from_u32(Self::EQ_TYPE);
+
+    /// A non-nullable reference to a struct aka `(ref struct)`.
+    pub const REF_STRUCT: Self = RefType::from_u32(Self::STRUCT_TYPE);
+
+    /// A non-nullable reference to an array aka `(ref array)`.
+    pub const REF_ARRAY: Self = RefType::from_u32(Self::ARRAY_TYPE);
+
+    /// A non-nullable reference to an i31 object aka `(ref i31)`.
+    pub const REF_I31: Self = RefType::from_u32(Self::I31_TYPE);
```

Comment:
While I realize this is following the naming of the text format, `I31REF` vs `REF_I31` feels a little too subtle to me as the only distinguishing factor being the nullability. I'm not sure of better names myself, though, so could this perhaps see how long we can go without the `REF_*` names instead?

Reply:
I've added these new constants because it was awkward to use `RefType::new()`, which returns `Option<RefType>` and then pulling the actual RefType out of it. 

I'm more than happy to rename the constants introduced in a prior PR with explicit NULL, e.g. `REF_NULL_ANY` instead of `ANYREF` (modeled by the shorthand names, which IMHO simply introduce keywords without new meaning).

Reply:
Also, more than happy to replace printed shorthand reftypes with `(ref null <heap_type>)` throughout.

Reply:
Bikeshed: how about `RefType::I31` for a non-nullable i31 and `RefType::I31.nullable()` for a nullable i31? This'd remove the `RefType::*REF` types like `RefType::ANYREF` and would be replaced with `RefType::ANY.nullable()` for example.

Reply:
Love this idea 👍

Reply:
🤔 some of the (older) `*REF`s (like `FUNCREF` and `EXTERNREF`) are already used in pattern matches and other stuff a bit.

Let me see if I can safely remove the newer ones keeping the older ones intact.

Reply:
Turns out, one can't simply use `RefType::ANY.nullable()` as a pattern match, even if `nullable()` is a `const fn`!

Reply:
Fixed!